pkg/amiclean/ami_cleaner.go

```diff
@@ -0,0 +1,149 @@
+// AMIClean defines parameters for cleaning up AMIs based on the Branch and
+// Expiration Date.
+type AMIClean struct {
+	Delete         bool
+	Branch         string
+	Invert         bool
+	ExpirationDate time.Time
+	Logger         *zap.Logger
+	EC2Client      *ec2.EC2
+}
+
+// GetImages gets us all the private AMIs on our account so that they can be
+// looked through later. We have to do this here because the AWS API does not
+// allow you to search for AMIs by creation date or by *not* having a tag set to
+// a certain value, which would speed this up considerably.
+func (a *AMIClean) GetImages() (*ec2.DescribeImagesOutput, error) {
+	var output *ec2.DescribeImagesOutput
+
+	input := &ec2.DescribeImagesInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("is-public"),
+				Values: []*string{aws.String("false")},
+			},
+		},
+	}
+
+	output, err := a.EC2Client.DescribeImages(input)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return output, err
+}
+
+// FindImagesToPurge looks through the AMIs available and produces a slice of
+// ec2.Image objects to put on the chopping block based on the contents of the
+// AMIClean struct.
+func (a *AMIClean) FindImagesToPurge(output *ec2.DescribeImagesOutput) []*ec2.Image {
+	var ImagesToPurge []*ec2.Image
+	for _, image := range output.Images {
+		imageCreationTime, _ := time.Parse(RFC8601, *image.CreationDate)
+		if imageCreationTime.After(a.ExpirationDate) {
+			continue
+		} else {
+			if a.Invert {
+				for _, tag := range image.Tags {
+					if *tag.Key == "Branch" && *tag.Value != a.Branch {
+						a.Logger.Info("selected ami for purging",
+							zap.String("ami-id",
+								*image.ImageId),
+							zap.String("ami-branch-tag",
+								*tag.Value),
+							zap.String("ami-creation-date",
+								imageCreationTime.String()),
+						)
+						ImagesToPurge =
+							append(ImagesToPurge, image)
+					}
+				}
+			} else {
+				for _, tag := range image.Tags {
+					if *tag.Key == "Branch" && *tag.Value == a.Branch {
+						a.Logger.Info("selected ami for purging",
+							zap.String("ami-id",
+								*image.ImageId),
+							zap.String("ami-branch-tag",
+								*tag.Value),
+							zap.String("ami-creation-date",
+								imageCreationTime.String()),
+						)
+						ImagesToPurge =
+							append(ImagesToPurge, image)
+					}
+				}
+			}
+		}
+	}
+	return ImagesToPurge
+}
+
+// PurgeImages takes a slice of ec2.Image objects and processes them in turn,
+// deregistering their AMI ID and then deleting their snapshot IDs. We want to
+// do this image by image because if we have an error, we don't want a ton of
+// orphaned snapshots lying around.
+func (a *AMIClean) PurgeImages(images []*ec2.Image) error {
+	for _, image := range images {
+		// There may be multiple snapshots attached to a single AMI,
+		// so we need to build a list and iterate on them.
+		var snapshotIds []*string
+		for _, blockDevice := range image.BlockDeviceMappings {
+			snapshotID := *blockDevice.Ebs.SnapshotId
```

Comment:
Do we have complete assurance that at this point, there will always be a value for `blockDevice.Ebs.SnapshotId`?

Reply:
Anything that uses a EBS block device for the AMI will have this -- you *can* have AMIs with instance storage instead, in which case this won't work (and there's S3 resources you need to clean up instead). I figured this would be good enough for an MVP, but maybe some guard rails are in order.

Reply:
My most recent commit adds an additional guard rail for this, checking to see if the AMI has an ebs root device; if it doesn't it will log a failure and continue.